File: recipes/subms-spsc-ring-buffer/java/src/main/java/com/submillisecond/recipes/spsc/SampleApp.java
package com.submillisecond.recipes.spsc;

import com.submillisecond.recipes.spsc.features.Bulk;
import com.submillisecond.recipes.spsc.features.Metrics;
import com.submillisecond.recipes.spsc.features.MpmcDisruptor;
import com.submillisecond.recipes.spsc.features.MpscFanIn;
import com.submillisecond.recipes.spsc.features.WaitStrategies;

import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Sample app: a tour of {@code subms-spsc-ring-buffer}, base API first, then each
 * optional feature. Run:
 * {@code mvn -q compile exec:java -Dexec.mainClass=com.submillisecond.recipes.spsc.SampleApp}
 *
 * <p>Domain: a market-data feed handler thread hands ticks (element = the tick's
 * sequence number) to a strategy thread over the wait-free ring. A full ring
 * makes the feed handler drop the tick (backpressure) rather than block.
 *
 * <ul>
 *   <li>base            - feed-handler -&gt; strategy handoff, drop-on-full
 *   <li>bulk            - drain a NIC receive batch in one fenced call
 *   <li>wait-strategies - a blocking handoff when a wakeup cost is affordable
 *   <li>mpsc-fan-in     - many venue feeds into one strategy, still wait-free per feed
 *   <li>mpmc-disruptor  - broadcast one tick stream to strategy + risk monitor
 *   <li>metrics         - per-instance enqueue/dequeue + max-depth counters
 * </ul>
 */
public final class SampleApp {

    public static void main(String[] args) throws InterruptedException {
        baseFeedToStrategy();
        bulkBatchIngest();
        blockingHandoff();
        manyVenueFanIn();
        broadcastToStrategyAndRisk();
        instrumentedHandoff();
    }

    /** Base API: a feed-handler thread pushes ticks to a strategy thread. Both
     * ends are wait-free; a full ring makes {@code tryPush} return false so the
     * feed handler drops the tick instead of stalling the hot path. */
    static void baseFeedToStrategy() throws InterruptedException {
        System.out.println("== base: feed-handler -> strategy handoff ==");

        // Drop-on-full is the caller's decision, shown on a small ring:
        // capacity 4, six ticks offered, the last two are dropped.
        SpscRingBuffer<Long> small = new SpscRingBuffer<>(4);
        SpscRingBuffer<Long>.Producer stx = small.producer();
        int dropped = 0;
        for (long seq = 0; seq < 6; seq++) {
            if (!stx.tryPush(seq)) dropped++;
        }
        System.out.println("  cap-4 ring, 6 offered -> " + dropped + " dropped under backpressure");
        if (dropped != 2) throw new AssertionError("two ticks past capacity are dropped, not blocked");

        // Occupancy is what a queue-depth alarm reads, and peek lets the strategy
        // inspect the oldest tick before deciding to consume it.
        SpscRingBuffer<Long>.Consumer srx = small.consumer();
        System.out.println("  depth " + srx.size() + "/" + small.capacity()
            + " full=" + stx.isFull() + ", oldest queued seq " + srx.peek());
        System.out.println("  dropped " + srx.clear() + " stale ticks on resync");

        // Steady state: a drained ring loses nothing and preserves feed order.
        long n = 50_000;
        SpscRingBuffer<Long> ring = new SpscRingBuffer<>(1024);
        SpscRingBuffer<Long>.Producer tx = ring.producer();
        SpscRingBuffer<Long>.Consumer rx = ring.consumer();

        Thread feed = new Thread(() -> {
            for (long seq = 0; seq < n; seq++) {
                while (!tx.tryPush(seq)) Thread.onSpinWait();
            }
        });
        AtomicLong received = new AtomicLong();
        Thread strategy = new Thread(() -> {
            long expected = 0;
            while (expected < n) {
                Long tick = rx.tryPop();
                if (tick != null) {
                    if (tick != expected) throw new AssertionError("ticks arrive in feed order");
                    expected++;
                }
            }
            received.set(expected);
        });
        feed.start();
        strategy.start();
        feed.join();
        strategy.join();
        System.out.println("  streamed " + received.get() + " ticks in order, zero loss when drained");
        if (received.get() != n) throw new AssertionError("every tick received");
    }

    /** bulk: a feed handler lifts a whole batch off one NIC receive.
     * {@code tryEnqueueBulk} copies the run in behind a single release fence
     * instead of one per tick; the strategy drains behind one acquire fence. */
    static void bulkBatchIngest() {
        System.out.println("\n== bulk: batch a NIC receive into the ring ==");
        SpscRingBuffer<Long> ring = new SpscRingBuffer<>(16);
        SpscRingBuffer<Long>.Producer tx = ring.producer();
        SpscRingBuffer<Long>.Consumer rx = ring.consumer();

        Long[] batch = new Long[10];
        for (int i = 0; i < batch.length; i++) batch[i] = 20_000L + i;

        int pushed = Bulk.tryEnqueueBulk(tx, batch);
        System.out.println("  offered " + batch.length + " ticks, took " + pushed + " in one fenced call");
        if (pushed != 10) throw new AssertionError("full batch fits");

        Long[] out = new Long[10];
        int drained = Bulk.tryDequeueBulk(rx, out);
        System.out.println("  drained " + drained + " in one fenced call");
        if (drained != 10) throw new AssertionError("full batch drains");
        for (int i = 0; i < out.length; i++) {
            if (!out[i].equals(batch[i])) throw new AssertionError("bulk preserves order");
        }
    }

    /** wait-strategies: wrap the non-blocking ends in a blocking handle with a
     * backoff policy. {@code YieldStrategy} lets other threads run between
     * retries, a sensible default when feed and strategy share cores. */
    static void blockingHandoff() throws InterruptedException {
        System.out.println("\n== wait-strategies: blocking handoff (yield backoff) ==");
        SpscRingBuffer<Long> ring = new SpscRingBuffer<>(8);
        WaitStrategies.BlockingSpscProducer<Long> producer =
                new WaitStrategies.BlockingSpscProducer<>(ring.producer(), new WaitStrategies.YieldStrategy());
        WaitStrategies.BlockingSpscConsumer<Long> consumer =
                new WaitStrategies.BlockingSpscConsumer<>(ring.consumer(), new WaitStrategies.YieldStrategy());

        long n = 5_000;
        Thread feed = new Thread(() -> {
            for (long seq = 0; seq < n; seq++) producer.push(seq);
        });
        Thread strategy = new Thread(() -> {
            for (long expected = 0; expected < n; expected++) {
                long tick = consumer.pop();
                if (tick != expected) throw new AssertionError("blocking pop keeps feed order");
            }
        });
        feed.start();
        strategy.start();
        feed.join();
        strategy.join();
        System.out.println("  handed off " + n + " ticks, producer blocked on full instead of dropping");
    }

    /** mpsc-fan-in: several venue feeds, one strategy consumer. Each feed owns
     * an independent SPSC ring (still wait-free against its own counter); the
     * consumer round-robins so a quiet venue never starves a busy one. */
    static void manyVenueFanIn() throws InterruptedException {
        System.out.println("\n== mpsc-fan-in: three venue feeds -> one strategy ==");
        int venues = 3;
        long perVenue = 20_000;
        MpscFanIn<Long> fanIn = new MpscFanIn<>(venues, 256);
        MpscFanIn<Long>.Consumer consumer = fanIn.consumer();

        Thread[] feeds = new Thread[venues];
        for (int v = 0; v < venues; v++) {
            MpscFanIn<Long>.Producer p = fanIn.producer(v);
            feeds[v] = new Thread(() -> {
                for (long seq = 0; seq < perVenue; seq++) {
                    while (!p.tryPush(seq)) Thread.onSpinWait();
                }
            });
        }
        long total = perVenue * venues;
        AtomicLong got = new AtomicLong();
        Thread strategy = new Thread(() -> {
            long local = 0;
            while (local < total) {
                if (consumer.tryPop() != null) local++;
            }
            got.set(local);
        });
        for (Thread f : feeds) f.start();
        strategy.start();
        for (Thread f : feeds) f.join();
        strategy.join();
        System.out.println("  " + venues + " feeds x " + perVenue + " ticks -> consumer drained " + got.get());
        if (got.get() != total) throw new AssertionError("every venue drained");
    }

    /** mpmc-disruptor: broadcast one tick stream to independent readers, a
     * strategy and a risk monitor, each of which sees every published tick.
     * (For each tick handled by exactly one reader, use mpsc-fan-in.) */
    static void broadcastToStrategyAndRisk() {
        System.out.println("\n== mpmc-disruptor: broadcast to strategy + risk ==");
        long n = 8;
        MpmcDisruptor<Long> disruptor = new MpmcDisruptor<>(16, 2);
        MpmcDisruptor<Long>.Producer producer = disruptor.producer();
        List<MpmcDisruptor<Long>.Consumer> consumers = disruptor.consumers();
        MpmcDisruptor<Long>.Consumer strategy = consumers.get(0);
        MpmcDisruptor<Long>.Consumer risk = consumers.get(1);

        long published = 0;
        long stratSeen = 0;
        long riskSeen = 0;
        while (published < n) {
            while (published < n && producer.tryPublish(published)) published++;
            Long v;
            while ((v = strategy.tryConsume()) != null) {
                if (v != stratSeen) throw new AssertionError("strategy sees every tick in order");
                stratSeen++;
            }
            while ((v = risk.tryConsume()) != null) {
                if (v != riskSeen) throw new AssertionError("risk sees every tick in order");
                riskSeen++;
            }
        }
        System.out.println("  published " + published + "; strategy saw " + stratSeen + ", risk saw " + riskSeen);
        if (stratSeen != n || riskSeen != n) throw new AssertionError("both readers see every tick");
    }

    /** metrics: wrap a base pair to count enqueue/dequeue success + fail and
     * track the high-water depth, at the cost of one atomic increment per op. */
    static void instrumentedHandoff() {
        System.out.println("\n== metrics: instrumented feed handoff ==");
        SpscRingBuffer<Long> ring = new SpscRingBuffer<>(4);
        Metrics.Instrumented<Long> inst = new Metrics.Instrumented<>(ring.producer(), ring.consumer());

        for (long seq = 0; seq < 4; seq++) {
            if (!inst.producer.tryPush(seq)) throw new AssertionError("first 4 fit");
        }
        if (inst.producer.tryPush(99L)) throw new AssertionError("5th is dropped");
        for (int i = 0; i < 4; i++) {
            if (inst.consumer.tryPop() == null) throw new AssertionError("4 drain");
        }
        if (inst.consumer.tryPop() != null) throw new AssertionError("ring empty");

        Metrics.RingMetricsSnapshot snap = inst.metrics.snapshot();
        System.out.println("  enqueued " + snap.enqueueSuccess + " (dropped " + snap.enqueueFail
                + "), dequeued " + snap.dequeueSuccess + " (empty " + snap.dequeueFail
                + "), peak depth " + snap.maxDepthObserved);
        if (snap.enqueueSuccess != 4 || snap.enqueueFail != 1) throw new AssertionError("enqueue counts");
        if (snap.dequeueSuccess != 4 || snap.dequeueFail != 1) throw new AssertionError("dequeue counts");
        if (snap.maxDepthObserved != 4) throw new AssertionError("peak depth");
    }

    private SampleApp() {}
}

File: recipes/subms-spsc-ring-buffer/rust/examples/sample_app.rs
//! Sample app: a tour of `subms-spsc-ring-buffer`, base API first, then each
//! optional feature. Run the base with `cargo run --example sample_app`; add
//! `--all-features` (or a subset like `--features bulk`) to see the feature
//! sections light up.
//!
//! Domain: a market-data feed handler thread hands ticks to a strategy thread
//! over the wait-free ring. A full ring makes the feed handler drop the tick
//! (backpressure) rather than block the hot path.
//!
//! * base            - feed-handler -> strategy handoff, drop-on-full
//! * bulk            - drain a NIC receive batch in one fenced call
//! * wait-strategies - a blocking handoff when a wakeup cost is affordable
//! * mpsc-fan-in     - many venue feeds into one strategy, still wait-free per feed
//! * mpmc-disruptor  - broadcast one tick stream to strategy + risk monitor
//! * metrics         - per-instance enqueue/dequeue + max-depth counters

use std::thread;

use subms_spsc_ring_buffer::SpscRingBuffer;

#[derive(Clone, Copy, Debug, PartialEq)]
struct Tick {
    seq: u64,
    price_cents: u32,
}

fn main() {
    base_feed_to_strategy();

    #[cfg(feature = "bulk")]
    bulk_batch_ingest();

    #[cfg(feature = "wait-strategies")]
    blocking_handoff();

    #[cfg(feature = "mpsc-fan-in")]
    many_venue_fan_in();

    #[cfg(feature = "mpmc-disruptor")]
    broadcast_to_strategy_and_risk();

    #[cfg(feature = "metrics")]
    instrumented_handoff();
}

/// Base API: a feed-handler thread pushes ticks to a strategy thread. Both ends
/// are wait-free; when the ring fills, `try_push` hands the tick back so the
/// feed handler drops it instead of stalling the hot path.
fn base_feed_to_strategy() {
    println!("== base: feed-handler -> strategy handoff ==");

    // Drop-on-full is the caller's decision. Shown deterministically on a
    // small ring: capacity 4, six ticks offered, the last two are dropped.
    let (mut tx, mut rx) = SpscRingBuffer::with_capacity::<Tick>(4);
    let mut dropped = 0usize;
    for seq in 0..6u64 {
        let tick = Tick {
            seq,
            price_cents: 10_000 + seq as u32,
        };
        if tx.try_push(tick).is_err() {
            dropped += 1;
        }
    }
    println!("  cap-4 ring, 6 offered -> {dropped} dropped under backpressure");
    assert_eq!(
        dropped, 2,
        "two ticks past capacity are dropped, not blocked"
    );

    // Occupancy is what a queue-depth alarm reads, and peek lets the strategy
    // inspect the oldest tick before deciding to consume it.
    let oldest = rx.peek().expect("ring is full").seq;
    println!(
        "  depth {}/{} full={}, oldest queued seq {oldest}",
        rx.len(),
        rx.capacity(),
        tx.is_full()
    );
    println!("  dropped {} stale ticks on resync", rx.clear());

    // Steady state: a drained ring loses nothing and preserves feed order.
    let n = 50_000u64;
    let (mut tx, mut rx) = SpscRingBuffer::with_capacity::<Tick>(1024);
    let feed = thread::spawn(move || {
        for seq in 0..n {
            let tick = Tick {
                seq,
                price_cents: 10_000 + (seq % 500) as u32,
            };
            while tx.try_push(tick).is_err() {
                std::hint::spin_loop();
            }
        }
    });
    let strategy = thread::spawn(move || {
        let mut expected = 0u64;
        while expected < n {
            if let Some(tick) = rx.try_pop() {
                assert_eq!(tick.seq, expected, "ticks arrive in feed order");
                expected += 1;
            }
        }
        expected
    });
    feed.join().unwrap();
    let received = strategy.join().unwrap();
    println!("  streamed {received} ticks in order, zero loss when drained");
    assert_eq!(received, n);
}

/// `bulk` feature: a feed handler often lifts a whole batch of ticks off one
/// NIC receive. `try_enqueue_bulk` copies the run in behind a single release
/// fence instead of one per tick; the strategy drains behind one acquire fence.
#[cfg(feature = "bulk")]
fn bulk_batch_ingest() {
    println!("\n== bulk: batch a NIC receive into the ring ==");
    let (mut tx, mut rx) = SpscRingBuffer::with_capacity::<Tick>(16);
    let batch: Vec<Tick> = (0..10)
        .map(|seq| Tick {
            seq,
            price_cents: 20_000 + seq as u32,
        })
        .collect();

    let pushed = tx.try_enqueue_bulk(&batch);
    println!(
        "  offered {} ticks, took {pushed} in one fenced call",
        batch.len()
    );
    assert_eq!(pushed, 10);

    let mut out = [Tick {
        seq: 0,
        price_cents: 0,
    }; 10];
    let drained = rx.try_dequeue_bulk(&mut out);
    println!("  drained {drained} in one fenced call");
    assert_eq!(drained, 10);
    assert_eq!(out.as_slice(), batch.as_slice(), "bulk preserves order");
}

/// `wait-strategies` feature: wrap the non-blocking ends in a blocking handle
/// with a backoff policy. `YieldStrategy` lets other threads run between
/// retries, a sensible default when the feed and strategy share cores.
#[cfg(feature = "wait-strategies")]
fn blocking_handoff() {
    use subms_spsc_ring_buffer::{BlockingSpscConsumer, BlockingSpscProducer, YieldStrategy};

    println!("\n== wait-strategies: blocking handoff (yield backoff) ==");
    let (tx, rx) = SpscRingBuffer::with_capacity::<Tick>(8);
    let mut producer = BlockingSpscProducer::new(tx, YieldStrategy);
    let mut consumer = BlockingSpscConsumer::new(rx, YieldStrategy);

    let n = 5_000u64;
    let feed = thread::spawn(move || {
        for seq in 0..n {
            producer.push(Tick {
                seq,
                price_cents: 30_000,
            });
        }
    });
    let strategy = thread::spawn(move || {
        for expected in 0..n {
            let tick = consumer.pop();
            assert_eq!(tick.seq, expected, "blocking pop keeps feed order");
        }
        n
    });
    feed.join().unwrap();
    let got = strategy.join().unwrap();
    println!("  handed off {got} ticks, producer blocked on full instead of dropping");
    assert_eq!(got, n);
}

/// `mpsc-fan-in` feature: several venue feeds, one strategy consumer. Each feed
/// owns an independent SPSC ring (still wait-free against its own counter); the
/// consumer round-robins so a quiet venue never starves a busy one.
#[cfg(feature = "mpsc-fan-in")]
fn many_venue_fan_in() {
    use subms_spsc_ring_buffer::MpscFanIn;

    println!("\n== mpsc-fan-in: three venue feeds -> one strategy ==");
    let venues = 3usize;
    let per_venue = 20_000u64;
    let (mut producers, mut consumer) = MpscFanIn::with_capacity::<Tick>(venues, 256);

    let mut feeds = Vec::new();
    for venue in 0..venues {
        let mut p = producers.remove(0);
        feeds.push(thread::spawn(move || {
            for seq in 0..per_venue {
                let tick = Tick {
                    seq,
                    price_cents: 40_000 + venue as u32,
                };
                while p.try_push(tick).is_err() {
                    std::hint::spin_loop();
                }
            }
        }));
    }
    let total = per_venue * venues as u64;
    let strategy = thread::spawn(move || {
        let mut got = 0u64;
        while got < total {
            if consumer.try_pop().is_some() {
                got += 1;
            }
        }
        got
    });
    for f in feeds {
        f.join().unwrap();
    }
    let got = strategy.join().unwrap();
    println!("  {venues} feeds x {per_venue} ticks -> consumer drained {got}");
    assert_eq!(got, total);
}

/// `mpmc-disruptor` feature: broadcast one tick stream to independent readers,
/// a strategy and a risk monitor, each of which sees every published tick. (To
/// have each tick handled by exactly one reader, reach for `mpsc-fan-in`.)
#[cfg(feature = "mpmc-disruptor")]
fn broadcast_to_strategy_and_risk() {
    use subms_spsc_ring_buffer::MpmcDisruptor;

    println!("\n== mpmc-disruptor: broadcast to strategy + risk ==");
    let n = 8u64;
    let (producer, mut consumers) = MpmcDisruptor::with_consumers::<Tick>(16, 2);
    let (strategy, rest) = consumers.split_at_mut(1);
    let strategy = &mut strategy[0];
    let risk = &mut rest[0];

    // Small and single-threaded so the tour self-verifies; the threaded
    // broadcast path is pinned in the tests.
    let mut published = 0u64;
    let mut strat_seen = Vec::new();
    let mut risk_seen = Vec::new();
    while published < n {
        while published < n
            && producer
                .try_publish(Tick {
                    seq: published,
                    price_cents: 50_000,
                })
                .is_ok()
        {
            published += 1;
        }
        while let Some(t) = strategy.try_consume() {
            strat_seen.push(t.seq);
        }
        while let Some(t) = risk.try_consume() {
            risk_seen.push(t.seq);
        }
    }
    println!(
        "  published {published}; strategy saw {}, risk saw {}",
        strat_seen.len(),
        risk_seen.len()
    );
    let expected: Vec<u64> = (0..n).collect();
    assert_eq!(strat_seen, expected, "strategy sees every tick");
    assert_eq!(risk_seen, expected, "risk monitor sees every tick too");
}

/// `metrics` feature: wrap a base pair to count enqueue/dequeue success + fail
/// and track the high-water depth - operational stats for a feed, at the cost
/// of one atomic increment per op.
#[cfg(feature = "metrics")]
fn instrumented_handoff() {
    use subms_spsc_ring_buffer::InstrumentedSpsc;

    println!("\n== metrics: instrumented feed handoff ==");
    let (tx, rx) = SpscRingBuffer::with_capacity::<Tick>(4);
    let (mut tx, mut rx, metrics) = InstrumentedSpsc::wrap(tx, rx);

    // Fill to capacity, one overflow drop, then drain plus one empty poll.
    for seq in 0..4u64 {
        tx.try_push(Tick {
            seq,
            price_cents: 60_000,
        })
        .unwrap();
    }
    assert!(
        tx.try_push(Tick {
            seq: 99,
            price_cents: 0
        })
        .is_err()
    );
    for _ in 0..4 {
        rx.try_pop().unwrap();
    }
    assert!(rx.try_pop().is_none());

    let snap = metrics.snapshot();
    println!(
        "  enqueued {} (dropped {}), dequeued {} (empty {}), peak depth {}",
        snap.enqueue_success,
        snap.enqueue_fail,
        snap.dequeue_success,
        snap.dequeue_fail,
        snap.max_depth_observed
    );
    assert_eq!(snap.enqueue_success, 4);
    assert_eq!(snap.enqueue_fail, 1);
    assert_eq!(snap.dequeue_success, 4);
    assert_eq!(snap.dequeue_fail, 1);
    assert_eq!(snap.max_depth_observed, 4);
}
